pkg/validate/openshift_cluster.go

```diff
@@ -0,0 +1,216 @@
+package validate
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/Azure/acs-engine/pkg/api/osa/vlabs"
+)
+
+var rxRfc1123 = regexp.MustCompile(`(?i)^` +
+	`([a-z0-9]|[a-z0-9][-a-z0-9]{0,61}[a-z0-9])` +
+	`(\.([a-z0-9]|[a-z0-9][-a-z0-9]{0,61}[a-z0-9]))*` +
+	`$`)
+
+var rxAgentPoolProfileVNetSubnetID = regexp.MustCompile(`(?i)^` +
+	`/subscriptions/[^/]+` +
+	`/resourceGroups/[^/]+` +
+	`/providers/Microsoft\.Network` +
+	`/virtualNetworks/[^/]+` +
+	`/subnets/[^/]+` +
+	`$`)
+
+var validAgentPoolProfileNames = map[string]struct{}{
+	string(vlabs.AgentPoolProfileRoleCompute): struct{}{},
+	string(vlabs.AgentPoolProfileRoleInfra):   struct{}{},
+	string(vlabs.AgentPoolProfileRoleMaster):  struct{}{},
+}
+
+var validRouterProfileNames = map[string]struct{}{
+	"default": struct{}{},
+}
+
+func isValidHostname(h string) bool {
+	return len(h) <= 255 && rxRfc1123.MatchString(h)
+}
+
+// OpenShiftCluster validates an OpenShiftCluster struct
+func OpenShiftCluster(oc *vlabs.OpenShiftCluster) (errs []error) {
+	if oc.Location == "" {
+		errs = append(errs, fmt.Errorf("invalid location %q", oc.Location))
+	}
+
+	if oc.Name == "" {
+		errs = append(errs, fmt.Errorf("invalid name %q", oc.Name))
+	}
+
+	if oc.Properties != nil {
+		errs = append(errs, validateProperties(oc.Properties)...)
+	}
+
+	return
+}
+
+func validateProperties(p *vlabs.Properties) (errs []error) {
+	switch p.ProvisioningState {
+	case "",
+		vlabs.Creating,
+		vlabs.Updating,
+		vlabs.Failed,
+		vlabs.Succeeded,
+		vlabs.Deleting,
+		vlabs.Migrating,
+		vlabs.Upgrading:
+	default:
+		errs = append(errs, fmt.Errorf("invalid properties.provisioningState %q", p.ProvisioningState))
+	}
+
+	switch p.OpenShiftVersion {
+	case "v3.10":
+	default:
+		errs = append(errs, fmt.Errorf("invalid properties.openShiftVersion %q", p.OpenShiftVersion))
+	}
+
+	if p.PublicHostname != "" && !isValidHostname(p.PublicHostname) {
+		errs = append(errs, fmt.Errorf("invalid properties.publicHostname %q", p.PublicHostname))
+	}
+
+	if p.FQDN != "" && !isValidHostname(p.FQDN) {
+		errs = append(errs, fmt.Errorf("invalid properties.fqdn %q", p.FQDN))
+	}
+
+	errs = append(errs, validateRouterProfiles(p.RouterProfiles)...)
+
+	errs = append(errs, validateAgentPoolProfiles(p.AgentPoolProfiles)...)
+
+	errs = append(errs, validateServicePrincipalProfile(&p.ServicePrincipalProfile)...)
+
+	return
+}
+
+func validateRouterProfiles(rps []vlabs.RouterProfile) (errs []error) {
+	rpmap := map[string]vlabs.RouterProfile{}
+
+	for _, rp := range rps {
+		if _, found := rpmap[rp.Name]; found {
+			errs = append(errs, fmt.Errorf("duplicate properties.routerProfiles %q", rp.Name))
+		}
+		rpmap[rp.Name] = rp
+
+		errs = append(errs, validateRouterProfile(rp)...)
+	}
+
+	for name := range validRouterProfileNames {
+		if _, found := rpmap[name]; !found {
+			errs = append(errs, fmt.Errorf("invalid properties.routerProfiles[%q]", name))
```

Comment:
The error message can probably be clearer that they are missing a valid router profile.